lib/cylc/task_events_mgr.py

```diff
@@ -1048,3 +1054,63 @@ def _reset_job_timers(self, itask):
         LOG.info('[%s] -%s', itask, message)
         # Set next poll time
         self.check_poll_time(itask)
+
+    @staticmethod
+    def process_execution_polling_delays(
+        delays, time_limit, time_limit_delays
+    ):
+        """Create list of intervals after starting at which to poll a task.
+
+        Args:
+            delays: Input is Execution Polling intervals.
+            time_limit_delays: Execution Time Limit Polling Intervals.
+            time_limit: Execution Time Limit.
+
+        Returns: List of delays from start of task.
+
+        Examples:
+
+            >>> this = TaskEventsManager.process_execution_polling_delays
+
+            # Basic example:
+            >>> this([40, 35], 100, [10])
+            [40, 35, 35, 10]
+
+            # Second 40 second delay gets lopped off the list because it's
+            # after the execution time limit:
+            >>> this([40, 40], 60, [10])
+            [40, 30, 10]
+
+            # Expand last item in exection polling intervals to fill the
+            # execution time limit:
+            >>> this([5, 20], 100, [10])
+            [5, 20, 20, 20, 20, 25, 10]
+
+            # There are no execution polling intervals set - polling starts
+            # at execution time limit:
+            >>> this([], 10, [5])
+            [15, 5]
+
+            # We have a list of execution time limit polling intervals,
+            >>> this([10], 25, [5, 6, 7, 8])
+            [10, 10, 10, 6, 7, 8]
+        """
+        if sum(delays) > time_limit:
+            # Remove excessive polling before time limit
+            while sum(delays) > time_limit:
+                del delays[-1]
+        elif delays:
+            # But fill up the gap before time limit
+            size = int((time_limit - sum(delays)) / delays[-1])
+            delays.extend([delays[-1]] * size)
+
+        # After the last delay before the execution time limit add the
+        # delay to get to the execution_time_limit
+        if len(time_limit_delays) > 1:
+            time_limit_delays[0] += time_limit - sum(delays)
+        else:
+            delays.append(
+                time_limit_delays[0] + time_limit - sum(delays))
```

Comment:
In the Cylc 8 PR this was simplified to
```suggestion
        if len(time_limit_delays) == 1:
            time_limit_delays.append(time_limit_delays[0])
        time_limit_delays[0] += time_limit - sum(delays)
```